Bump of libgd2, libwebp and libvpx, and rehaul of make-rules for better sub-component support

jimklimov opened the pull request:
Recently we mentioned libvpx... and there were a few more repos in the project that were not stagnant, not for Windows, and low-hanging fruit to build. So why not? ;)

jimklimov committed:
Bumped libvpx-1.4.0 to version 1.6.0

pyhalov commented:
We already have components/library/libwebp

jimklimov commented:
uh, ok - I see it is there, but older (0.4.0)... I think this PR will grow with multi-SO packaging like other recent library upgrades. VPX also wants a bump...

pyhalov commented on components/library/webm/libvpx/Makefile:
```diff
@@ -0,0 +1,62 @@
+#
```
Why do we want moving them to webm?

jimklimov replied:
They are part of the same project family. OI-Userland tends to group those (e.g. samba of various versions, docbook-related stuff, etc.)

It would also make the directory structure maybe easier to navigate (less entries in each dir) if we choose to keep several complete versions of the package delivered (e.g. not only .so files, but also includes, etc.)

xen0l replied:
Moving them to specific directory under category should not be desired. With fixed structure, we know the maximum depth where components can be found. It also copies behaviour of FreeBSD ports or pkgsrc.

pyhalov commented on components/library/webm/libwebm/Makefile:
```diff
@@ -0,0 +1,76 @@
+			/usr/bin/elfedit -e 'dyn:value -s  RUNPATH '"$(USRLIBDIR)" $$F ; \
+		done ; \
+		$(CP) libwebm.so $(PROTOUSRLIBDIR) && \
+		$(CP) sample sample_muxer vttdemux webm2pes $(PROTOUSRBINDIR)
```
$(TOUCH) $@ ?

jimklimov replied:
thanks

pyhalov commented on components/library/webm/libwebm/Makefile:
```diff
@@ -0,0 +1,76 @@
+			/usr/bin/elfedit -e 'dyn:value -s  RUNPATH '"$(USRLIBDIR64)" $$F ; \
+		done ; \
+		$(CP) libwebm.so $(PROTOUSRLIBDIR64) && \
+		$(CP) sample sample_muxer vttdemux webm2pes $(PROTOUSRBINDIR64)
```
$(TOUCH) $@ ?

pyhalov commented on components/library/webm/libwebm/Makefile:
```diff
@@ -0,0 +1,76 @@
+	$(MKDIR) $(PROTOUSRBINDIR) $(PROTOUSRLIBDIR)
+	cd $(BUILD_DIR_32) && \
+		for F in libwebm.so sample sample_muxer vttdemux webm2pes ; do \
+			/usr/bin/elfedit -e "dyn:rpath ''" $$F && \
```
Isn't it better to path out -Wl,-R or is it non-trivial?

pyhalov commented on components/library/webm/libwebm/Makefile:
```diff
@@ -42,6 +42,7 @@ TARGET.32 = x86-solaris-gcc
 TARGET.64 = x86_64-solaris-gcc
 
 CXXFLAGS += -std=gnu++11
+LDFLAGS += -Wl,-R
```
How does it help???

jimklimov replied:
I think it does not... I just copied your suggestion. From the few samples
in repo, seems there should be optional(?) comma and some libpath. No idea
what to do or why, especially why it is mis-generated in the first place.

## 

Typos courtesy of GMail on my Samsung Android
31.05.2016 17:37 пользователь "Alexander Pyhalov" notifications@github.com
написал:

> In components/library/webm/libwebm/Makefile
> https://github.com/OpenIndiana/oi-userland/pull/2067#discussion_r65207153
> :
> 
> > @@ -42,6 +42,7 @@ TARGET.32 = x86-solaris-gcc
> >  TARGET.64 = x86_64-solaris-gcc
> > 
> >  CXXFLAGS += -std=gnu++11
> > +LDFLAGS += -Wl,-R
> 
> How does it help???
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> https://github.com/OpenIndiana/oi-userland/pull/2067/files/0fec5f9329b63b3f5aee75977b0ed4e1bc9aed4f..31a4570fd74ad6c7d6cde3aee8b796653cd69551#r65207153,
> or mute the thread
> https://github.com/notifications/unsubscribe/ABj5lCCGx8OsRur89OafGM6IVzFSvXt-ks5qHFWzgaJpZM4IpMZj
> .

pyhalov replied:
I thought about patching makefiles...

jimklimov replied:
Hopefully got it working now... at least 'gmake clean publish' succeeds.

xen0l commented:
I did some REQUIRED_PACKAGES changes and touched libvpx, too. This caused conflict with this PR for which I would like to apologize.

jimklimov commented:
Fixed mergeability back :)

xen0l commented:
Please, don't introduce categories within categories. Otherwise, it's good to he merged. @pyhalov what do you think?

pyhalov commented on components/library/webm/libwebm/Makefile:
```diff
@@ -0,0 +1,74 @@
+install: $(INSTALL_32_and_64)
+
+BUILT_LIBS = libwebm.so
+BUILT_BINS = sample sample_muxer vttdemux webm2pes
```
can we move samples to /usr/demo/libwebm ?

jimklimov replied:
done
from cursory reading, "sample" says it is a "sample application" while others seem to be actual tools using the library

pyhalov commented on components/library/webm/libwebm/libwebm.p5m:
```diff
@@ -0,0 +1,36 @@
+
+license $(COMPONENT_LICENSE_FILE) license='$(COMPONENT_LICENSE)'
+
+file path=usr/bin/$(MACH64)/sample
```
Name is too generic, if it's example, it's better to be moved under /usr/demo/libwebm

pyhalov commented:
Have you checked that webkitgtk2 still builds fine after this update?

jimklimov commented:
Point taken about demo dir - though i'll be away for about 10 days... No
tests done about webkit or other consumers.

## 

Typos courtesy of GMail on my Samsung Android
30.06.2016 21:40 пользователь "Alexander Pyhalov" notifications@github.com
написал:

> Have you checked that webkitgtk2 still builds fine after this update?
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> https://github.com/OpenIndiana/oi-userland/pull/2067#issuecomment-229766522,
> or mute the thread
> https://github.com/notifications/unsubscribe/ABj5lEuTAUIqTOhhpCr2o9NmLHtTxu_1ks5qRBu1gaJpZM4IpMZj
> .

jimklimov commented:
Moved back to common `libraries/` category, and updated some versions (keeping old binary .so's distributed earlier). Can't rebase in my repo for some reason (e-too-many-force-pushes?) so feel free to squash and merge via guthub interface ;p

pyhalov commented on components/library/libwebm/Makefile:
```diff
@@ -0,0 +1,74 @@
+		$(CP) $(BUILT_LIBS) $(PROTOUSRLIBDIR64) && \
+		$(CP) $(BUILT_BINS) $(PROTOUSRBINDIR64)
+	$(TOUCH) $@
+
```
I don't understand this... Does libwebm ship any headers?

jimklimov replied:
I'll check later. Probably should...

## 

Typos courtesy of GMail on my Samsung Android

17.07.2016 8:01 пользователь "Alexander Pyhalov" notifications@github.com
написал:

> In components/library/libwebm/Makefile
> https://github.com/OpenIndiana/oi-userland/pull/2067#discussion_r71077087
> :
> 
> > +BUILT_BINS = sample sample_muxer vttdemux webm2pes
> > +
> > +$(BUILD_DIR_32)/.installed: $(BUILD_DIR_32)/.built
> > -   $(MKDIR) $(PROTOUSRBINDIR) $(PROTOUSRLIBDIR)
> > -   cd $(BUILD_DIR_32) && \
> > -       $(CP) $(BUILT_LIBS) $(PROTOUSRLIBDIR) && \
> > -       $(CP) $(BUILT_BINS) $(PROTOUSRBINDIR)
> > -   $(TOUCH) $@
> >   +
> >   +$(BUILD_DIR_64)/.installed: $(BUILD_DIR_64)/.built
> > -   $(MKDIR) $(PROTOUSRBINDIR64) $(PROTOUSRLIBDIR64)
> > -   cd $(BUILD_DIR_64) && \
> > -       $(CP) $(BUILT_LIBS) $(PROTOUSRLIBDIR64) && \
> > -       $(CP) $(BUILT_BINS) $(PROTOUSRBINDIR64)
> > -   $(TOUCH) $@
> >   +
> 
> I don't understand this... Does libwebm ship any headers?
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> https://github.com/OpenIndiana/oi-userland/pull/2067/files/e0c1e02fe3eb35c4cd0e11711697db0f49a6394d#r71077087,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/ABj5lAM3SqQXOq5f1x_cq3IHqOkZig_kks5qWcUwgaJpZM4IpMZj
> .

jimklimov replied:
Strangely enough, it does not. There are no headers in the source dir prepared by CMake, and just 4 headers accompanying some .cpp files in the original tarball.

I guess this is used internally somehow.

pyhalov commented on components/library/libwebp/libwebp.p5m:
```diff
@@ -1,40 +1,121 @@
+
+
+# Include files are also a bit different for two releases, so we provide both
+file path=usr/include/webp-0.4.0/decode.h
```
Do we really need ship headers from older webp?

jimklimov replied:
I'm away from computer so memory may be vague, but IIRC the two releases
were not fully compatible in function signatures, so there may be software
ready to work with one and requiring chiseling for another.

Maybe it was memory about another library though.

## 

Typos courtesy of GMail on my Samsung Android

17.07.2016 8:03 пользователь "Alexander Pyhalov" notifications@github.com
написал:

> In components/library/libwebp/libwebp.p5m
> https://github.com/OpenIndiana/oi-userland/pull/2067#discussion_r71077128
> :
> 
> >  link path=usr/lib/libwebp.so.5 target=libwebp.so.5.0.0
> >  file path=usr/lib/libwebp.so.5.0.0
> > -file path=usr/lib/pkgconfig/libwebp.pc
> > -file path=usr/share/man/man1/cwebp.1
> > -file path=usr/share/man/man1/dwebp.1
> > +
> > +file path=usr/lib/$(MACH64)/pkgconfig/libwebp-0.4.0.pc
> > +file path=usr/lib/pkgconfig/libwebp-0.4.0.pc
> > +
> > +
> > +# Include files are also a bit different for two releases, so we provide both
> > +file path=usr/include/webp-0.4.0/decode.h
> 
> Do we really need ship headers from older webp?
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> https://github.com/OpenIndiana/oi-userland/pull/2067/files/e0c1e02fe3eb35c4cd0e11711697db0f49a6394d#r71077128,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/ABj5lLaOXV1aaevjVahD1sKip5wtIRkPks5qWcWWgaJpZM4IpMZj
> .

jimklimov replied:
I was wrong, the headers are pretty compatible. One notable difference is in a structure of YUV colorspaces supported. Also many earlier reserved+padded entries are now renamed and used apparently - but in a size-compatible manner.

And there are many more headers in the new version :)

pyhalov commented:
The components, which depend on libvpx:
  pkg:/image/gnuplot
  pkg:/library/audio/gstreamer/plugin/bad
  pkg:/library/gd
  pkg:/media/vlc
  pkg:/video/ffmpeg
on libwebp:
  pkg:/image/imagemagick
  pkg:/library/desktop/webkitgtk2

It's not a great list. Perhaps, let's just bump these libraries, rebuild dependencies and ensure that everything is fine?

jimklimov committed:
libwebp bumped to 0.5.1 (still deliver SONAMEs for 0.4.0 delivered earlier); Updated REQUIRED_PACKAGES

jimklimov commented:
Hopefully squashed well and still intact ;p

Verified that building (gmake clean publish) and installing all 3 webm libs as well as gd2 component on top of them - works. BTW, GD2 also has seen some development and security fixes recently... ;)

jimklimov committed:
libvpx Makefile : greatly simplified thanks to improvements in make-rules, and dropped delivery of 1.5.0 version that was never released in repo

pyhalov commented on components/library/gd2/Makefile:
```diff
@@ -79,23 +100,86 @@ CONFIGURE_OPTIONS +=	--disable-rpath
+# so we work around it with much complexity below... and still only get one
+# set of *.ln files built so far. Also they may be useless in practice and
+# just get dropped in a later round of recipying.
+include $(WS_MAKE_RULES)/lint-libraries.mk
```
I'd rather just removed lint libraries, as gd is not used by illumos-gate build and removed additional complexity.

pyhalov commented on components/library/libvpx/Makefile:
```diff
@@ -11,45 +11,174 @@
+SOURCE_DIR_2=			$(COMPONENT_DIR)/$(COMPONENT_SRC_2)
+
+# Another intermediate version
+COMPONENT_VERSION_3=	1.5.0
```
We don't need 1.5.0 binary, we have never delivered it

jimklimov replied:
I kept it just in case some components (new TB?) would specifically work
with 1.5.0 and are not yet compatible with 1.6.0. Can remove :)

## 

Typos courtesy of GMail on my Samsung Android

26.07.2016 20:27 пользователь "Alexander Pyhalov" notifications@github.com
написал:

> In components/library/libvpx/Makefile
> https://github.com/OpenIndiana/oi-userland/pull/2067#discussion_r72308652
> :
> 
> > +COMPONENT_ARCHIVE_URL_1= \
> > -   https://github.com/webmproject/$(COMPONENT_NAME)/archive/v$(COMPONENT_VERSION).tar.gz
> >   +SOURCE_DIR_1=          $(COMPONENT_DIR)/$(COMPONENT_SRC_1)
> >   +
> >   +# Older version for legacy compatibility
> >   +COMPONENT_VERSION_2    = 1.4.0
> >   +COMPONENT_ARCHIVE_HASH_2= \
> > -   sha256:eca30ea7fae954286c9fe9de9d377128f36b56ea6b8691427783b20c67bcfc13
> >   +COMPONENT_ARCHIVE_URL_2= \
> > -   https://github.com/webmproject/$(COMPONENT_NAME)/archive/v$(COMPONENT_VERSION_2).tar.gz
> >   +COMPONENT_SRC_2=       $(COMPONENT_NAME)-$(COMPONENT_VERSION_2)
> >   +COMPONENT_ARCHIVE_2=   $(COMPONENT_SRC_2).tar.gz
> >   +SOURCE_DIR_2=          $(COMPONENT_DIR)/$(COMPONENT_SRC_2)
> >   +
> >   +# Another intermediate version
> >   +COMPONENT_VERSION_3=   1.5.0
> 
> We don't need 1.5.0 binary, we have never delivered it
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> https://github.com/OpenIndiana/oi-userland/pull/2067/files/fc74bf1be1058091903ec605cbd9d40e7a55885a#r72308652,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/ABj5lLQZH0qhhsLZ3kSaXSf0km2YCD9Qks5qZlFwgaJpZM4IpMZj
> .

pyhalov replied:
We don't deliver compilation symlinks or headers, so it's useless. I'd removed it.

jimklimov committed:
prep-unpack.mk : shield from misuse when a COMPONENT_SRC$(1) is not set (e.g. a recipe with subcomponents only defines the suffixed _N variables)

jimklimov committed:
lint-libraries.mk : rectify the linting rules so they work with sub-components (multiple sources) too

jimklimov committed:
prep.mk : automate configure, build, install and package support for sub-components (when multiple sources are involved) so ultimate recipes can be simplified; also ensure ordering of build steps such as requiring that prep happens ;)

jimklimov commented:
Cleaning up the mess (poor make-rules automation support for components with several sources, and none for linting such components) allowed to improve the core makefiles and simplify the libvpx and libgd2 recipes a lot.

While testing with random other packages, also found and fixed a bug with zlib recipe (it is linted, so was of interest here). Whether we would ultimately decide to drop lint recipes or not, at this point they work properly now.

Also verified for regression-testing vs. some other recipes with sub-projects, e.g. docbook/\* - they also build and publish okay with these make-rules changes.

Bumping the libgd to 2.2.3 may have to wait (patches should be developed or snatched to make this version build), but the recipe will be based on this advanced foundation - so I suggest to merge this PR before it rots and requires tinkering and delays to become mergeable again :)

jimklimov committed:
libwebp Makefile : greatly simplified thanks to improvements in make-rules

pyhalov commented on make-rules/prep-unpack.mk:
```diff
@@ -34,6 +34,7 @@ PCK_SUFFIXES = $(subst COMPONENT_ARCHIVE_,, \
 define unpack-rules
 ifdef COMPONENT_ARCHIVE$(1)
 ifdef COMPONENT_SRC$(1)
+ifneq ($(strip $$(COMPONENT_SRC$(1))),)
```
why do you need to change this?

pyhalov commented on make-rules/prep.mk:
```diff
@@ -27,21 +27,285 @@ include $(WS_MAKE_RULES)/prep-download.mk
+
+ifneq ($(strip $(1)),)
+
+ifdef COMPONENT_VERSION$(1)
```
Why do we need this in prep.mk?

pyhalov commented:
It seems almost sane, but I'd like one more man to look through make-rules/*.mk changes. @xenol, what do you think?

jimklimov commented:
I guess there would be a slight merge-freeze due the later set of make-rules changes. Re-reading the gmake mans, it seems that "ifeq" is invalid in the "define"d function context and should be converted to "$(if ...)" functions. And those behave funny in my experiments - not the way I expect from docs and blogs. So expect a PR to rectify that...

jimklimov commented:
Merge-freeze state deasserted.

xen0l commented:
I see the logic in these changes, but I think they are too invasive :|

jimklimov commented:
FYI : extracted the bits on ips.mk into a separate PR #2538

jimklimov committed:
Add libgd-2.2.3 recipe and still deliver the older version

jimklimov commented:
FYI : extracted the bits on better support for recipes with sub-components into PR #2583 
The unrelated fix on zlib posted to PR #2584 
Introduction of libwebm to PR #2585

jimklimov commented:
PR updated vs. current upstream and heads of respective branches split-away for other PRs (see above).

jimklimov commented:
FYI : For now I chose to remove the libwebm commits from remains of PR #2067 which was started to integrate it and update related things ;) This is because of uncertainty whether at all, and in what scope, the webm project should be packaged rather than built-in and brought along by its consumer projects.

Integration of libwebm is tracked separately, in PR #2585 and does not really depend on other commits in this branch.

alarcher commented:
@jimklimov Sorry I did not notice your PR. Are you able to rebase after I merge libwebp?

jimklimov commented:
@alarcher I guess so. Better late than never ;)

_Sent from my Xiaomi Redmi Note 4 using [FastHub](https://play.google.com/store/apps/details?id=com.fastaccess.github)_
